File: sol/1745.py
from functools import cache

class Solution:
    def checkPartitioning(self, s: str) -> bool:
        @cache
        def isPal(i, j):
            if i >= j:
                return True
            return s[i] == s[j] and isPal(i + 1, j - 1)
        n = len(s)
        b = tuple(i for i in range(n) if isPal(0, i))
        e = tuple(i for i in range(n) if isPal(i, n - 1))
        return any(isPal(i + 1, j - 1) for i in b for j in e if i <= j - 2)

File: sol/1745/1745.cpp
struct Solution {
    bool checkPartitioning(string &s) {
        int n = s.size();
        vector<vector<bool>> isPal(n, vector<bool>(n,  false));
        for (int i = 0; i < n; ++i){
            isPal[i][i] = true;
        }
        for (int i = 0; i < n - 1; ++i){
            isPal[i][i + 1] = s[i] == s[i + 1];
        }
        for (int l = 3; l <= n; ++l){
            for (int i = 0, j = l - 1; j < n; ++i, ++j){
                isPal[i][j] = (s[i] == s[j]) && isPal[i + 1][j - 1];
            }
        }
        for (int i = 0; i < n - 2; ++i){
            if (!isPal[0][i]){
                continue;
            }
            for (int j = i + 1; j < n - 1; ++j){
                if (isPal[i + 1][j] && isPal[j + 1][n - 1]){
                    return true;
                }
            }
        }
        return false;
    }
};
